persistent/Database/Persist/Sql/Run.hs

```diff
@@ -105,28 +107,63 @@ runSqlPoolWithHooks r pconn i before after onException =
 
 -- | This function is how 'runSqlPoolWithHooks' is defined.
 --
--- It's currently the most general function for using a SQL pool.
---
--- @since 2.13.0.0
+-- @since 2.13.3.0
 runSqlPoolWithExtensibleHooks
     :: forall backend m a. (MonadUnliftIO m, BackendCompatible SqlBackend backend)
     => ReaderT backend m a
     -> Pool backend
     -> Maybe IsolationLevel
     -> SqlPoolHooks m backend
     -> m a
-runSqlPoolWithExtensibleHooks r pconn i SqlPoolHooks{..} =
-    withRunInIO $ \runInIO ->
-    withResource pconn $ \conn ->
-    UE.mask $ \restore -> do
-        conn' <- restore $ runInIO $ alterBackend conn
-        _ <- restore $ runInIO $ runBefore conn' i
-        a <- restore (runInIO (runReaderT r conn'))
-            `UE.catchAny` \e -> do
-                _ <- restore $ runInIO $ runOnException conn' i e
-                UE.throwIO e
-        _ <- restore $ runInIO $ runAfter conn' i
-        pure a
+runSqlPoolWithExtensibleHooks = runSqlPoolWithExtensibleHooksRetry $ const False
+
+-- | This function is equivalent to 'runSqlPoolWithExtensibleHooks' but
+-- additionally allows specifying an exception predicate. On encountering an
+-- exception during a transaction, this predicate decides whether or not the
+-- transaction should be retried. This can be used to build various retrying
+-- schemes, such as retrying on serialization/deadlock errors when running
+-- transactions at serializable isolation level.
+--
+-- Note that even though the predicate operates on 'UE.SomeException', it is
+-- only applied to synchronous exceptions. Asynchronous exceptions are always
+-- rethrown and will never trigger a retry of the transaction.
+--
+-- Considering @persistent@ abstracts over specific SQL backends, you will
+-- likely need to reach for a backend-specific exception type when defining an
+-- exception predicate.
+--
+-- @since 2.14.6.0
+runSqlPoolWithExtensibleHooksRetry
+    :: forall backend m a. (MonadUnliftIO m, BackendCompatible SqlBackend backend)
+    => (UE.SomeException -> Bool)
```

Comment:
It's worth pointing out that the API has been intentionally kept simple here: so long as exceptions match the predicate, the transaction will always be retried. In the case of serialization failures at repeatable-read and serializable isolation specifically, the recommendation from the PostgreSQL docs and elsewhere is to retry these failing transactions unconditionally.

That guidance is what drove this simpler API, but users may have [other failures](https://www.postgresql.org/docs/current/mvcc-serialization-failure-handling.html) (e.g. uniqueness violations) they would like to only retry up to a fixed number of times or retry based on some more sophisticated policy. The simple exception predicate approach in this PR would not work for those more complicated cases. Considering the more complicated cases are rare, it seemed prudent to keep the retry API as simple as possible for now. However, depending upon the exception predicate the user specifies, they may get themselves into a situation where `persistent` indefinitely retries. For these cases, they could wrap their `runSqlPoolWithExtensibleHooksRetry` call in a `timeout`.